### PR TITLE
fix: issue #926

### LINKED ISSUE #926: Renaming a metabolite compartment does not update the reaction compartment
#### Problem description

Had some problems in micom and finally pinpointed them to some issue with compartment names in cobrapy. I wanted to rename compartments for metabolites in a model and noted that this did not propagate to reactions.

#### Code Sample

Create a [minimal, complete, verifiable example](https://stackoverflow.com/help/mcve).

```python
rxn = Reaction("r1")
met = Metabolite("m1", compartment="c")
rxn.add_metabolites({met: 2})

print(rxn.compartments)

met.compartment = "e"

print(rxn.compartments)
```

Funnily, if you would not run the first print it would actually show the correct compartment. This is due to caching of the compartment in Reaction which infers it from metabolites once but never updates.

#### Actual Output

```
{"c"}
{"c"}
```

#### Expected Output

```
{"c"}
{"e"}
```

#### Dependency Information

Please paste the output of `python -c "from cobra import show_versions;
show_versions()"` in the details block below.

<details>

System Information
==================
OS         Darwin
OS-release 18.7.0
Python      3.6.7

Package Versions
================
cobra                                    0.17.1
depinfo                                   1.5.1
future                                   0.18.2
numpy                                    1.17.3
optlang                                   1.4.4
pandas                                   0.25.3
pip                                      19.3.1
python-libsbml-experimental                   ?
ruamel.yaml                              0.16.5
setuptools                  41.6.0.post20191101
six                                      1.13.0
swiglpk                                  4.65.0
wheel                                    0.33.6

</details>

#### Proposed solution

Remove the compartment caching and have only metabolites store compartment info. Reactions infer them from their metabolites only.

### PR DESCRIPTION
**Pull request by dtmvandenberg**
* [x] fix #926 
* [x] description of feature/fix
* [x] tests added/passed
* [x] add an entry to the [next release](../release-notes/next-release.md)

### CODE REVIEW

File: src/cobra/core/reaction.py [hunk 1]
```diff
@@ -999,13 +996,11 @@ def check_mass_balance(self):
...
-        if self._compartments is None:
-            self._compartments = {
-                met.compartment
-                for met in self._metabolites
-                if met.compartment is not None
+        return {
+            met.compartment
+            for met in self._metabolites
+            if met.compartment is not None
             }
```
Comment (cdiener) [inline src/cobra/core/reaction.py]: Formatting is a bit off.

```suggestion
        return {
            met.compartment
            for met in self._metabolites
            if met.compartment is not None
        }
```

**Review by cdiener — CHANGES_REQUESTED** [review]
Hi, looks good. A small formatting change so it passes the pep8 checks. We require the code to be pep8-compliant. If you're using vscode or something you can usually activate flake8 checks which helps with that.

Also there is a `.swp` file that probably was not supposed to be added so this would need to be removed as well.

And finally, it would be great if you could add a test so this does not pop up again. For that you can add something along the lines of:

```python
def test_compartment_changes(model):
    rxn = model.reactions.EX_ac_e
    assert rxn.reactants[0].compartment in rxn.compartments
    rxn.reactants[0].compartment = "blub"
    assert rxn.reactants[0].compartment in rxn.compartments
```

to the end of `src/cobra/test/test_core/test_core_reaction.py`. 

Oh yeah, and add a short note in the `release-notes/next-release.md`.

**Comment by dtmvandenberg** [comment]
@cdiener I changed the files you requested and i commited them but Github will not update so I am stuck...

File: src/cobra/core/reaction.py [hunk 2]
```diff
@@ -999,13 +996,11 @@ def check_mass_balance(self):
...
-                met.compartment
-                for met in self._metabolites
-                if met.compartment is not None
-            }
-        return self._compartments
+        return {
+            met.compartment
+            for met in self._metabolites
+            if met.compartment is not None
+        }
```
Comment (cdiener) [inline src/cobra/core/reaction.py]: Sorry, my bad. This is what the formatter wants now.

```suggestion
        return {
            met.compartment for met in self._metabolites if met.compartment is not None
        }
```

**Review by cdiener — CHANGES_REQUESTED** [review]
:+1:

File: .gitignore [hunk 3]
```diff
@@ -302,3 +302,4 @@ gurobi.log
 /documentation_builder/test*\.*
 /.benchmarks
 /.testmondata
+.vscode/settings.json
```
Comment (cdiener) [inline .gitignore]: Even `.vscode/*`.

File: src/cobra/core/reaction.py [hunk 4]
```diff
@@ -999,13 +996,11 @@ def check_mass_balance(self):
...
     def compartments(self):
         """lists compartments the metabolites are in"""
-        if self._compartments is None:
-            self._compartments = {
-                met.compartment
-                for met in self._metabolites
-                if met.compartment is not None
-            }
-        return self._compartments
+        return {
```
Comment (cdiener) [inline src/cobra/core/reaction.py:999]: This is the old formatting again.

**Review by cdiener — CHANGES_REQUESTED** [review]
Some of the old commits crept back in it seems (the `.swp` file is back as well). Pretty close though.

**Comment by dtmvandenberg** [comment]
@cdiener Everything should be okay now, sorry for the amount of work. Do you know how to initialize a new codecov report? Then everything should be in order for the merge.

**Review by cdiener — APPROVED** [review]
Looks good now. Thanks for contributing!

**Pull request merged**